### ISSUE TITLE
[api-extractor] Internal Error: fetchReferencedAstSymbol() called for a source file that was not analyzed

### ISSUE THREAD

**Issue by natalieethell** [Microsoft/web-build-tools#1054]
I'm getting the following error when trying to run api-extractor 7.0.13:

```
Internal Error: fetchReferencedAstSymbol() called for a source file that was not analyzed
```

You should be able to reproduce the error on this branch after running "rush install" and "rush build": https://github.com/natalieethell/office-ui-fabric-react/tree/apiExtractorError

**Comment by pgonzal** [comment]
The affected API is:

```ts
export { IStyleFunctionOrObject } from '@uifabric/merge-styles';
```

The bug is that API Extractor's [_analyzeChildTree()](https://github.com/Microsoft/web-build-tools/blob/a521df999d7b6e6604a086b54b2e4c3e40613722/apps/api-extractor/src/analyzer/AstSymbolTable.ts#L195) naively assumes that `symbolNode.getSourceFile()` will refer to a source file that was already analyzed:

```ts
          const symbol: ts.Symbol | undefined = this._typeChecker.getSymbolAtLocation(symbolNode);
          if (!symbol) {
            throw new Error('Symbol not found for identifier: ' + symbolNode.getText());
          }

          const referencedAstSymbol: AstSymbol | undefined
            = this.fetchReferencedAstSymbol(symbol, symbolNode.getSourceFile());  <----
          if (referencedAstSymbol) {
            governingAstDeclaration._notifyReferencedAstSymbol(referencedAstSymbol);
          }
```

...but this is not true in general.

Here I think the edge case comes from this doozy of type algebra:

```ts
export declare type IStyleFunctionOrObject<TStylesProps, TStyleSet extends IStyleSet<TStyleSet>> = IStyleFunction<TStylesProps, TStyleSet> | Partial<TStyleSet>;
```

...where the `IStyleSet` generic type constraint recursively references its own declaration:

```ts
export declare type IStyleSet<TStyleSet extends IStyleSet<TStyleSet>> = {
    [P in keyof Omit<TStyleSet, 'subComponentStyles'>]: IStyle;
} & {
    subComponentStyles?: {
        [P in keyof TStyleSet['subComponentStyles']]: IStyleFunctionOrObject<any, IStyleSet<any>>;
    };
};
```

@cliffkoh sometimes I think you guys are intentionally trying to devise every pathological way to break API Extractor heheh. :-P

**Comment by cliffkoh** [comment]
Trust me, it was far from easy trying to correctly express merge-styles semantics :-P It was such a struggle to even figure out a type definition that TypeScript itself is happy with - devising ways to torture API was definitely not something I had any remaining bandwidth for :-)

**Comment by craigkovatch** [comment]
I'm seeing this error in 7.0.11 for one of our projects too (closed-source, ugh). But I'm not given any information about the symbol causing the error; running with `--debug` doesn't help either. @natalieethell how did you figure out which symbol was causing your trouble?

**Comment by octogonz** [comment]
We ran API Extractor against her Git branch using the VS Code debugger.  If you want to try that yourself, below are some steps.

However you would really want to be doing this using my PR branch from https://github.com/Microsoft/web-build-tools/pull/1077 since it changes the affected logic. In particular the `fetchReferencedAstSymbol()` error message is removed entirely.

### How to debug **API Extractor**

1. Make sure you're using Node 10, since its debugger support is way better
2. Build the **web-build-tools** repo (`rush install`, `rush rebuild`)
3. Open the **apps/api-extractor** folder in VS Code
4. Switch to the "Debug" tab (ctrl+shift=D) and then choose `"scratch"` from  list of configurations
5. Click the settings gear next to `"scratch"`, and edit this blurb to point the `"cwd"` (current working directory) at your project folder:

```json
    {
      "type": "node",
      "request": "launch",
      "name": "scratch",
      "program": "${workspaceFolder}/lib/start.js",
      "cwd": "(your project path)",
      "args": [
        "--debug",
        "run",
        "--local"
      ],
      "sourceMaps": true
    }
```

> Note that the `cwd` folder needs to have an **api-extractor.json** config file that is compatible with AE7.  You can find examples under the [build-tests](https://github.com/Microsoft/web-build-tools/tree/master/build-tests) folder, which all build using AE7.
6. Click the play button (F5) to start debugging. Exceptions are unfortunately caught by the command-line front end, so to trap them you will either need to search for the error message in the source code and set a breakpoint manually, or use VS Code's "Uncaught Exceptions" breakpoint (which generally you have to wait to enable until after all the unexceptional exceptions that NodeJS throws during startup).
7. When you trap the error in the debugger, you can walk upwards in the call stack and look for one of these types:

- `ts.Node`: Use `node.getSourceFile().fileName` to see the file path, and `node.getText()` to see the affected code (or for more context you can do `node.parent.getText()`, `node.parent.parent.getText()`, ...)
- `ts.Symbol`: Use `symbol.getDeclarations()[0]` to get the `ts.Node`
- `AstDeclaration`: use `astDeclaration.declaration` to get the `ts.Node`
- `AstSymbol`: Use `astSymbol.astDeclarations[0]` to get the `AstDeclaration`

**Comment by octogonz** [comment]
> However you would really want to be doing this using my PR branch from https://github.com/Microsoft/web-build-tools/pull/1077 since it changes the affected logic. In particular the `fetchReferencedAstSymbol()` error message is removed entirely.

@craigkovatch also just to be clear, PR https://github.com/Microsoft/web-build-tools/pull/1077 still has some bugs that I was working on (which you can see from the test outputs).  I'll remove the "**(not ready for review yet)**" suffix from the title once that's fixed.

**Comment by octogonz** [comment]
> I'll remove the "**(not ready for review yet)**" suffix from the title once that's fixed.

The PR is now ready, and I've validated that the OUIFR branch is now building successfully.  We'll merge it soon.

**Comment by octogonz** [comment]
This should be fixed in API Extractor 7.0.17 beta.  If not, please reopen this issue.

To install the beta, use this command:

```
npm install -g @microsoft/api-extractor@beta
```

(The "beta" releases have regular ordinary numbers like 7.0.17.  Our [NPM disttag](https://www.npmjs.com/package/@microsoft/api-extractor?activeTab=versions) prevents them from getting installed as the "latest" version if you simply run `npm install -g @microsoft/api-extractor`.  We will update the disttag as soon as version 7 is stable.)
